### PR TITLE
Utxoscriptindex

### PR DESCRIPTION
**Pull request by mgrychow**
Yet another attempt to introduce UTxO index by script. Based on #9806 with review remarks and new index framework introduced in #13033 taken into account.

### CODE REVIEW

**Comment by instagibbs** [comment]
Might be wortwhile to update it to using output descriptors, like `scantxoutset`: https://github.com/bitcoin/bitcoin/blob/master/src/rpc/blockchain.cpp#L2035

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -1079,6 +1082,145 @@ UniValue gettxout(const JSONRPCRequest& request)
...
+        if ((unsigned int)coin.nHeight != MEMPOOL_HEIGHT)
+            nConfirmations = pindex->nHeight - coin.nHeight + 1;
+        if (nConfirmations < minConf)
+            continue;
+
+        UniValue oScriptPubKey(UniValue::VOBJ);
+        ScriptPubKeyToUniv(coin.out.scriptPubKey, oScriptPubKey, true);
+
+        UniValue o(UniValue::VOBJ);
+        o.push_back(Pair("confirmations", nConfirmations));
```
[the remark this answers is not in the document]
Comment (mgrychow) [reply src/rpc/blockchain.cpp]: Corrected

**Comment by marcinja** [comment]
@mgrychow, I recently started working on an address-based index (not just for UTXOs) at #14053. I can improve my PR quite a bit using your changes to the ValidationInterface. If you separated that change (adding CBlockUndo's) into it's own commit this PR would be a bit cleaner, and I could use it for the address index :)

Anyway, will be reviewing this since it's pretty similar to the PR I just made.

**Comment by instagibbs** [comment]
@mgrychow Not going to bikeshed this PR to death, just noting it in case you've missed it.

File: src/index/utxoscriptindex.cpp [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+{
+    RegisterValidationInterface(this);
+}
+
+void UtxoScriptIndex::TransactionAddedToMempool(const CTransactionRef &ptxn) {
+    LOCK(cs_utxoCacheMempool);
+    const CTransaction &tx = *ptxn;
+    for (unsigned int i = 0; i < tx.vout.size(); i++)
+    {
+        if (tx.vout[i].IsNull() || tx.vout[i].scriptPubKey.IsUnspendable())
```
Comment (marcinja) [inline src/index/utxoscriptindex.cpp]: Is this check needed? In what case will a transaction be added to the mempool where this is true? I might be missing something.

File: src/index/utxoscriptindex.cpp [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+    }
+}
+
+void UtxoScriptIndex::removeUtxosOnTipDisconnected(const CBlock& block)
+{
+    if(block.vtx.size() == 0)
+        return;
+
+    unsigned int i = block.vtx.size() - 1;
+    while(true)
```
Comment (marcinja) [inline src/index/utxoscriptindex.cpp]: This could be a `do{ } while(i != 0)`

**Review by marcinja — CHANGES_REQUESTED** [review]

File: src/index/utxoscriptindex.cpp (hunk 3 above)
Comment (mgrychow) [reply src/index/utxoscriptindex.cpp]: Loop needs to execute for `i==0` too (that is block's tx with index 0); `i` is unsigned so after decrementation it overflows and such loop would be indefinite

**Comment by mgrychow** [comment]
Rebased

**Comment by MarcoFalke** [comment]
There hasn't been much activity lately and the patch still needs rebase, so I am closing this for now. Please let me know when you want to continue working on this, so the pull request can be re-opened.

**Pull request closed without merging**